pkg/userserver/user_server.go

```diff
@@ -0,0 +1,269 @@
+	}
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	targetURL, err := url.Parse(serviceProxyURL(tsc.Cluster))
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	tunnel, err := k.getTunnel(req.Context())
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(targetURL)
+	proxy.Transport = &http.Transport{
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			RootCAs:    serviceProxyRootCA,
+			MinVersion: tls.VersionTLS12,
+		},
+		// golang http pkg automaticly upgrade http connection to http2 connection, but http2 can not upgrade to SPDY which used in "kubectl exec".
+		// set ForceAttemptHTTP2 = false to prevent auto http2 upgration
+		ForceAttemptHTTP2: false,
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			klog.V(4).Infof("proxy dial to %s", addr)
+			// TODO: may find a way to cache the proxyConn.
+			return tunnel.DialContext(ctx, network, addr)
+		},
+	}
+
+	proxy.ErrorHandler = func(rw http.ResponseWriter, r *http.Request, e error) {
+		http.Error(rw, fmt.Sprintf("proxy to anp-proxy-server failed because %v", e), http.StatusBadGateway)
+		klog.Errorf("proxy to anp-proxy-server failed because %v", e)
+	}
+
+	klog.V(4).Infof("request scheme:%s; rawQuery:%s; path:%s", req.URL.Scheme, req.URL.RawQuery, req.URL.Path)
+
+	proxy.ServeHTTP(wr, utils.UpdateRequest(tsc, req))
+}
+
+func (k *userServer) Run(ctx context.Context) error {
+	var err error
+
+	klog.Info("begin to run user server")
+
+	if err = k.Validate(); err != nil {
+		klog.Fatal(err)
+	}
+
+	if err = k.init(ctx); err != nil {
+		klog.Fatal(err)
+	}
+
+	cc, err := addonutils.NewConfigChecker("user-server", k.proxyCACertPath, k.proxyCertPath, k.proxyKeyPath, k.serverCert, k.serverKey, k.serviceProxyCACertPath)
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	go func() {
+		if err = utils.ServeHealthProbes(":8000", cc.Check); err != nil {
+			klog.Fatal(err)
+		}
+	}()
+
+	klog.Infof("start https server on %d", k.serverPort)
+
+	s := &http.Server{
+		Addr:      fmt.Sprintf(":%d", k.serverPort),
+		TLSConfig: &tls.Config{MinVersion: tls.VersionTLS12},
+		Handler:   k,
+	}
+
+	err = s.ListenAndServeTLS(k.serverCert, k.serverKey)
+	if err != nil {
+		klog.Fatalf("failed to start user proxy server: %v", err)
+	}
+
+	return nil
+}
+
+// here use the same logic as in the cluster-proxy repo:
+// https://github.com/stolostron/cluster-proxy/blob/304b2ded6c1a651be9ba0f15af4edf1f65ac29df/pkg/proxyagent/agent/agent.go#L297
+func serviceProxyURL(clusterName string) string {
+	serviceProxyHost := fmt.Sprintf("cluster-%x", sha256.Sum256([]byte(clusterName)))[:64-len("cluster-")] + ".open-cluster-management.proxy"
```

Comment:
this sha256.Sum256([]byte(clusterName)))[:64-len("cluster-")] would need some comment. I actually think it should be in a separated line

Reply:
Extract this part in to a independent function in util package:
```
// Using SHA256 to hash cluster.name to:
// 1. Generate consistent and unique host names
// 2. Keep host name length under DNS limit (max 64 chars)
func GenerateServiceProxyHost(clusterName string) string {
	return fmt.Sprintf("cluster-%x",
		sha256.Sum256([]byte(clusterName)))[:64-len("cluster-")] + ".open-cluster-management.proxy"
}
```